pymc_extras/statespace/models/DFM.py

```diff
@@ -0,0 +1,811 @@
+            "error_cov",
+            "sigma_obs",
+            "beta",
+            "beta_sigma",
+        ]
+
+        # Handle cases where parameters should be excluded based on model settings
+        if self.factor_order == 0:
+            names.remove("factor_ar")
+        if self.error_order == 0:
+            names.remove("error_ar")
+        if self.error_cov_type in ["scalar", "diagonal"]:
+            names.remove("error_cov")
+        if self.error_cov_type == "unstructured":
+            names.remove("error_sigma")
+        if not self.measurement_error:
+            names.remove("sigma_obs")
+        if not self.exog_flag:
+            names.remove("beta")
+            names.remove("beta_sigma")
+        if self.exog_flag and not self.exog_innovations:
+            names.remove("beta_sigma")
+
+        return names
+
+    @property
+    def param_info(self) -> dict[str, dict[str, Any]]:
+        info = {
+            "x0": {
+                "shape": (self.k_states,),
+                "constraints": None,
+            },
+            "P0": {
+                "shape": (self.k_states, self.k_states),
+                "constraints": "Positive Semi-definite",
+            },
+            "factor_loadings": {
+                "shape": (self.k_endog, self.k_factors),
+                "constraints": None,
+            },
+            "factor_ar": {
+                "shape": (self.k_factors, self.factor_order * self.k_factors),
+                "constraints": None,
+            },
+            "error_ar": {
+                "shape": (
+                    self.k_endog,
+                    self.error_order * self.k_endog if self.error_var else self.error_order,
+                ),
+                "constraints": None,
+            },
+            "error_sigma": {
+                "shape": (self.k_endog,) if self.error_cov_type == "diagonal" else (),
+                "constraints": "Positive",
+            },
+            "error_cov": {
+                "shape": (self.k_endog, self.k_endog),
+                "constraints": "Positive Semi-definite",
+            },
+            "sigma_obs": {
+                "shape": (self.k_endog,),
+                "constraints": "Positive",
+            },
+            "beta": {
+                "shape": (self.k_exog_states,),
+                "constraints": None,
+            },
+            "beta_sigma": {
+                "shape": (self.k_exog_states,),
+                "constraints": "Positive",
+            },
+        }
+
+        for name in self.param_names:
+            info[name]["dims"] = self.param_dims[name]
+
+        return {name: info[name] for name in self.param_names}
+
+    @property
+    def state_names(self) -> list[str]:
+        """
+        Returns the names of the hidden states: first factor states (with lags),
+        idiosyncratic error states (with lags), then exogenous states.
+        """
+        names = []
+
+        for i in range(self.k_factors):
```

Comment:
Refactor these loops to be list comprehensions. Start with:

```py
names = [f'L{lag}.factor_{i}' for lag in range(max(self.factor_order ,1)) for i in range(k_factors)]
```

Then on subsequent loops you can use `names.extend([...])`, like you did for exog_flag